File: sortingAlgorithms/mergeSort.cpp
#include <bits/stdc++.h>
using namespace std;

void merge(vector<int>&arr, int start, int end, int mid){
    //find size of two portions
    int leftSize = mid-start+1;
    int rightSize = end-mid;

    //copy elements 
    vector<int> leftArr (leftSize);
    vector<int> rightArr (rightSize);

    for(int i=0; i<leftSize; i++){
        leftArr[i] = arr[start + i];
    }

    for(int i=0; i<rightSize; i++){
        rightArr[i] = arr[i+mid+1] ;
    }

    //merge two sorted array
    int i =0, j=0, k = start;

    while(i<leftSize && j<rightSize){
        if(leftArr[i]<rightArr[j]) arr[k++] =  leftArr[i++];
        else arr[k++] = rightArr[j++];
    }

    //push remaining elements 
    while(i<leftSize) arr[k++] = leftArr[i++];
    while(j<rightSize) arr[k++] = rightArr[j++];
}

void mergeSort(vector<int>& arr, int start, int end){
    if(start>=end) return;
    int mid = start+(end-start)/2;
    mergeSort(arr, start, mid);
    mergeSort(arr, mid+1, end);
    merge(arr, start, end, mid );
    
}

void printArr(const vector<int>& arr){
    for(int ele: arr){
        cout<<ele<<" ";
    }
}

int main(){
    vector<int> arr = {54,11,87,34};
    mergeSort(arr,0,arr.size()-1);
    printArr(arr);
}

File: sortingAlgorithms/MergeSort.java
package sortingAlgorithms;

public class MergeSort {
    static void merge(int[] arr, int start, int mid, int end) {
        // length of two sorted arrays
        int leftLen = mid - start + 1;
        int rightLen = end - mid;
        // create array
        int[] leftArr = new int[leftLen];
        int[] rightArr = new int[rightLen];
        // copy values
        for (int i = 0; i < leftLen; i++) {
            leftArr[i] = arr[start + i];
        }
        for (int i = 0; i < rightLen; i++) {
            rightArr[i] = arr[mid + 1 + i];
        }
        // merge
        int i = 0, j = 0;
        int k = start;
        while (i < leftLen && j < rightLen) {
            if (leftArr[i] <= rightArr[j])
                arr[k++] = leftArr[i++];
            else
                arr[k++] = rightArr[j++];
        }
        // fill the remaining
        while (i < leftLen)
            arr[k++] = leftArr[i++];
        while (j < rightLen)
            arr[k++] = rightArr[j++];
    }

    static void mergeSort(int[] arr, int start, int end) {
        if (start >= end) return;
        int mid = start + (end - start) / 2;
        mergeSort(arr, start, mid);
        mergeSort(arr, mid + 1, end);
        merge(arr, start, mid, end);

    }

    public static void main(String[] args) {
        int[] arr = { 3, 5, 7, 1, 9, 2, 4, 6, 8 };
        mergeSort(arr, 0, arr.length - 1);
        for (int ele : arr)
            System.out.print(ele + " ");
    }
}
